src/xref2/ref_tools.ml

```diff
@@ -589,51 +638,47 @@ and resolved2 (r, _) = resolved1 r
   let parent_sg = Tools.prefix_signature (parent_path, parent_sg) in
-  Find.any_in_sig parent_sg name >>= function
-  | `Module (_, _, m) ->
-      let name = ModuleName.of_string name in
+  find_ambiguous Find.any_in_sig parent_sg name >>= function
+  | `M (name, m) ->
       resolved3
-        (M.of_component env (Component.Delayed.get m)
+        (M.of_component env m
            (`Module (parent_path, name))
            (`Module (parent_ref, name)))
-  | `ModuleType (_, mt) ->
-      let name = ModuleTypeName.of_string name in
+  | `MT (name, mt) ->
       resolved3
-        (MT.of_component env (Component.Delayed.get mt)
+        (MT.of_component env mt
            (`ModuleType (parent_path, name))
            (`ModuleType (parent_ref, name)))
-  | `Type (_, _, t) ->
-      DT.of_component env (Component.Delayed.get t) ~parent_ref name
+  | `T (name, t) ->
+      DT.of_component env t ~parent_ref name
       >>= resolved2
-  | `Class (_, _, c) -> CL.of_component env c ~parent_ref name >>= resolved2
-  | `ClassType (_, _, ct) ->
-      CT.of_component env ct ~parent_ref name >>= resolved2
-  | `Value _ -> V.of_component env ~parent_ref name >>= resolved1
-  | `External _ -> V.external_of_component env ~parent_ref name >>= resolved1
-  | `Label _ -> L.of_component env ~parent_ref name >>= resolved1
-  | `Constructor (typ_name, _, _) ->
-      let parent = `Type (parent_ref, typ_name) in
-      CS.of_component env parent name >>= resolved1
-  | `Exception _ -> EX.of_component env ~parent_ref name >>= resolved1
-  | `ExtConstructor _ -> EC.of_component env ~parent_ref name >>= resolved1
-  | `Field (typ_name, _, _) ->
+  | `C (name, c) -> CL.of_component env c ~parent_ref name >>= resolved2
+  | `CT (name, ct) -> CT.of_component env ct ~parent_ref name >>= resolved2
+  | `V _ -> V.of_component env ~parent_ref name >>= resolved1
+  | `E _ -> V.external_of_component env ~parent_ref name >>= resolved1
+  | `L _ -> L.of_component env ~parent_ref name >>= resolved1
+  | `Exn _ -> EX.of_component env ~parent_ref name >>= resolved1
+  | `Ext _ -> EC.of_component env ~parent_ref name >>= resolved1
+  | `In_type (typ_name, _, r) -> (
       let parent = `Type (parent_ref, typ_name) in
-      F.of_component env parent name >>= resolved1
-  | `ModuleSubstitution _ | `Removed _ | `TypeSubstitution _ -> None
+      match r with
+      | `Cs _ -> CS.of_component env parent name >>= resolved1
+      | `Fd _ -> F.of_component env parent name >>= resolved1 )
+  | `Msub _ | `Tsub _ -> None
 
 let resolve_reference_dot_page env page name =
   L.in_page env page name >>= resolved1
 
 let resolve_reference_dot_type env ~parent_ref t name =
   Find.any_in_type t name >>= function
-  | `Constructor _ -> CS.of_component env parent_ref name >>= resolved1
-  | `Field _ -> F.of_component env parent_ref name >>= resolved1
+  | `Cs _ -> CS.of_component env parent_ref name >>= resolved1
+  | `Fd _ -> F.of_component env parent_ref name >>= resolved1
```

Comment:
I do find these slightly _too_ terse - I don't mind abbreviations generally but I don't like mentally tripping over them when I don't find them really obvious. Perhaps we could use `Constr` and `Fld`, `Meth` and `Instv`? `Mv` as `InstanceVariable` I find particularly baffling.

Reply:
I went a bit too far with abbreviations ! I used them to avoid confusion with paths/references/identifiers (and module name clash).
An alternative would be to use the full name with a prefix, eg. `F_constructor` in `Find` and `R_constructor` in `Ref_tools`. What do you think ?

Reply:
I changed to the `FConstructor` style.